app/celer-g4/RootIO.cc

```diff
@@ -62,76 +62,76 @@ HitRootIO::HitRootIO()
-        tree_.reset(new TTree(
-            "Events", "Hit collections", this->SplitLevel(), file_.get()));
+        tree_.reset(new TTree(this->TreeName(),
+                              this->TreeName(),
+                              this->SplitLevel(),
+                              file_.get()));
     }
 }
 
-//---------------------------------------------------------------------------//
-//! Default destructor
-HitRootIO::~HitRootIO() = default;
-
 //---------------------------------------------------------------------------//
 /*!
- * Return the static thread local singleton instance>
+ * Return the static thread local singleton instance.
  */
-HitRootIO* HitRootIO::Instance()
+RootIO* RootIO::Instance()
 {
-    static G4ThreadLocal HitRootIO instance;
+    static G4ThreadLocal RootIO instance;
     return &instance;
 }
 
 //---------------------------------------------------------------------------//
 /*!
- * Write sensitive hits to output in the form of HitRootEvent.
+ * Write sensitive hits to output in the form of EventData.
+ * See celeritas/io/EventData.hh
  */
-void HitRootIO::WriteHits(G4Event const* event)
+void RootIO::Write(G4Event const* event)
 {
     G4HCofThisEvent* hce = event->GetHCofThisEvent();
     if (hce == nullptr)
     {
         return;
     }
 
-    // Write the collection of sensitive hits into HitRootEvent
-    HitRootEvent hit_event;
-    hit_event.event_id = event->GetEventID();
+    // Write steps and collection of hits
+    EventData event_data;
+    event_data.event_id = event->GetEventID();
     for (int i = 0; i < hce->GetNumberOfCollections(); i++)
     {
+        std::vector<HitData> hits;
+
         G4VHitsCollection* hc = hce->GetHC(i);
-        std::string hcname = hc->GetName();
-        std::vector<G4VHit*> hits;
-        int number_of_hits = hc->GetSize();
-        for (int j = 0; j < number_of_hits; ++j)
+        for (std::size_t j = 0; j < hc->GetSize(); ++j)
         {
-            G4VHit* hit = hc->GetHit(j);
-            SensitiveHit* sd_hit = dynamic_cast<SensitiveHit*>(hit);
-            hits.push_back(sd_hit);
+            auto* sd_hit = dynamic_cast<SensitiveHit*>(hc->GetHit(j));
+            auto const& result = sd_hit->data();
+            event_data.steps.push_back(result.step);
+            hits.push_back(result.hit);
         }
-        hit_event.hcmap.insert(std::make_pair(hcname, std::move(hits)));
+        auto iter = detector_name_id_map_.find(hc->GetName());
+        CELER_ASSERT(iter != detector_name_id_map_.end());
+        event_data.hits.insert({iter->second, std::move(hits)});
     }
 
-    // Write a HitRootEvent into output ROOT file
-    this->WriteObject(&hit_event);
+    this->WriteObject(&event_data);
 }
 
 //---------------------------------------------------------------------------//
 /*!
- * Fill a HitRootEvent object.
+ * Fill event tree with event data.
  */
-void HitRootIO::WriteObject(HitRootEvent* hit_event)
+void RootIO::WriteObject(EventData* event_data)
```

Comment:
Just a reminder that the pattern in this function
```
SetAddress(...);
Fill();
ResetBranchAddresses();
```
is valid, if and only if there is NO other branches in the `TTree` as `TTree::Fill` will import the user data for ALL the branches.

Reply:
Yes! Ideally we do only store one branch, which has the event struct with all the data. This way reading the root file is as simple as possible: load events tree; set the address of the event branch; Get entries.

But this is a good point and I'll add a comment to the code.